extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,238 @@
+
+    /** @private {?Promise<!../../../src/service/xhr-impl.FetchResponse>} */
+    this.adPromise_ = null;
+
+    /** @private {Object<string, !./amp-ad-type-defs.Validator>} */
+    this.validators_ = map();
+
+    /** @private {Object<string, !./amp-ad-type-defs.Renderer>} */
+    this.renderers_ = map();
+
+    /** @private {Object<string, string>} */
+    this.recoveryModes_ = map();
+
+    /** @private {?./amp-ad-type-defs.LayoutInfoDef} */
+    this.initialSize_ = null;
+
+    /** @const @private {!Object} */
+    this.context_ = {
+      element,
+      win: this.win,
+    };
+
+    // Register default error modes.
+    for (const failureType in FailureType) {
+      this.recoveryModes_[failureType] = RecoveryModeType.COLLAPSE;
+    }
+
+    /**
+     * Number of times to retry a failed request. Zero by default.
+     * @type {number}
+     */
+    this.retryLimit_ = 0;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.FailureType} failure
+   * @param {!./amp-ad-type-defs.RecoveryModeType} recovery
+   * @final
+   */
+  onFailure(failure, recovery) {
+    if (this.recoveryModes_[failure]) {
+      dev().warn(TAG,
+          `Recovery mode for failure type ${failure} already registered!`);
+    }
+    this.recoveryModes_[failure] = recovery;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.Validator} validator
+   * @param {string=} type
+   * @final
+   */
+  registerValidator(validator, type = 'default') {
+    if (this.validators_[type]) {
+      dev().warn(TAG, `${type} validator already registered.`);
+    }
+    this.validators_[type] = validator;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.Renderer} renderer
+   * @param {string} type
+   * @final
+   */
+  registerRenderer(renderer, type) {
+    if (this.renderers_[type]) {
+      dev().warn(TAG, `Rendering mode already registered for type '${type}'`);
+    }
+    this.renderers_[type] = renderer;
+  }
+
+  /**
+   * @param {string} name
+   * @param {*} value
+   */
+  setContextField(name, value) {
+    this.context_[name] = value;
+  }
+
+  /**
+   * @return {string} The finalized ad request URL.
+   * @protected
+   * @abstract
+   */
+  getRequestUrl() {
+    // Subclass must override.
+  }
+
+  /** @param {number} retries */
+  setRequestRetries(retries) {
+    this.retryLimit_ = retries;
+  }
+
+  /**
+   * Processes the ad response as soon as the XHR request returns.
+   * @param {?../../../src/service/xhr-impl.FetchResponse} response
+   * @return {!Promise}
+   * @private
+   */
+  invokeValidator_(response) {
+    if (!response.arrayBuffer) {
+      return Promise.reject(this.handleFailure_(FailureType.INVALID_RESPONSE));
+    }
+    return response.arrayBuffer().then(unvalidatedBytes => {
+      const validatorType = response.headers.get('AMP-Ad-Response-Type')
+          || 'default';
+      dev().assert(this.validators_[validatorType],
+          'Validator never registered!');
+      try {
+        return this.validators_[validatorType].validate(
```

Comment:
try/catch will not work on promise chain.
do this instead:
```
return ...validate.(..)
    .catch(err => {
        return Promise.reject({type: FailureType.VALIDATOR_ERROR, details: err});
    });
```

Reply:
Done.